### PR TITLE
Allow annotated FlyteFile as task input argument

### PR DESCRIPTION
**Pull request by AdrianoKF**
# TL;DR

This PR allows annotated FlyteFiles to be used as input arguments of a task, e.g., when calculating a custom hash using `Annotated[FlyteFile, HashMethod(...)]`.

See the discussion on [Slack](https://flyte-org.slack.com/archives/CP2HDHKE1/p1683635974841289) for additional details.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Using an annotated FlyteFile type as an input to a task was previously impossible due to an exception being raised in `FlyteFilePathTransformer.to_python_value`.

This MR applies the fix previously used in `FlyteFilePathTransformer.to_literal` to permit using annotated FlyteFiles as either inputs or outputs of a task (see MR #1544), by unwrapping the origin type of the annotated type.

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3424

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/types/file/file.py [hunk 1]
```diff
@@ -26,6 +26,13 @@ def noop():
 T = typing.TypeVar("T")
 
 
+def _get_origin_type(t: Type) -> Type:
+    """Return the origin type for annotated types or the type itself, such that it can be used with ``issubclass()``"""
+    if get_origin(t) is Annotated:
+        return get_args(t)[0]
```
Comment (pingsutw) [inline flytekit/types/file/file.py:32]: we do this all the place in the `type_engine.py`. https://github.com/flyteorg/flytekit/blob/a1636a89daee2301a056b3eba41fe4dcea1257b7/flytekit/core/type_engine.py#L737-L748
could you move this function to `type_engine.py`, and import it instead.

**Review by pingsutw — COMMENTED** [review]
LGTM overall, thanks

File: flytekit/types/file/file.py (hunk 1 above)
Comment (AdrianoKF) [reply flytekit/types/file/file.py]: Moved the function there and refactored `type_engine.py` quite a bit in 093536fef. This should also make the handling of annotated types there more uniform (and added a few tests).

Is this what you had in mind?

**Review by pingsutw — COMMENTED** [review]
Thanks, LGTM. would you resolve branch conflict?

**Comment by AdrianoKF** [comment]
> Thanks, LGTM. would you resolve branch conflict?

Sorry, was sick for the past days, just merged upstream `master` and resolved the conflicts.

**Review by pingsutw — DISMISSED** [review]
@AdrianoKF hope you feel better.

**Comment by AdrianoKF** [comment]
Merged `master` again and fixed the test failures on Python 3.8 in 34600b081.

The remaining failing test on 3.11/Windows seems to be unrelated to my changes and I'm not familiar enough with that part of the code base to suggest a fix - anything I should do about that, @pingsutw?

**Review by pingsutw — APPROVED** [review]

**Comment by pingsutw** [comment]
@AdrianoKF never mind. That's flaky test.
